rust/sbp/src/parser/mod.rs

```diff
@@ -169,20 +170,114 @@ where
     }
 }
 
-pub(crate) fn read_string(buf: &mut &[u8]) -> Result<SbpString> {
-    let amount = buf.len();
-    let (head, tail) = buf.split_at(amount);
-    *buf = tail;
-    Ok(SbpString(head.to_vec()))
+pub(crate) trait SbpParse<T> {
+    fn parse(&mut self) -> Result<T>;
 }
 
-pub(crate) fn read_string_limit(buf: &mut &[u8], n: usize) -> Result<SbpString> {
-    let n = std::cmp::min(n, buf.len());
-    let (mut head, tail) = buf.split_at(n);
-    read_string(&mut head).map(|sbp_string| {
-        *buf = tail;
-        sbp_string
-    })
+impl SbpParse<u8> for &[u8] {
+    fn parse(&mut self) -> Result<u8> {
+        self.read_u8().map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<i8> for &[u8] {
+    fn parse(&mut self) -> Result<i8> {
+        self.read_i8().map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<u16> for &[u8] {
+    fn parse(&mut self) -> Result<u16> {
+        self.read_u16::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<i16> for &[u8] {
+    fn parse(&mut self) -> Result<i16> {
+        self.read_i16::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<u32> for &[u8] {
+    fn parse(&mut self) -> Result<u32> {
+        self.read_u32::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<i32> for &[u8] {
+    fn parse(&mut self) -> Result<i32> {
+        self.read_i32::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<u64> for &[u8] {
+    fn parse(&mut self) -> Result<u64> {
+        self.read_u64::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<i64> for &[u8] {
+    fn parse(&mut self) -> Result<i64> {
+        self.read_i64::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<f32> for &[u8] {
+    fn parse(&mut self) -> Result<f32> {
+        self.read_f32::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl SbpParse<f64> for &[u8] {
+    fn parse(&mut self) -> Result<f64> {
+        self.read_f64::<LittleEndian>()
+            .map_err(|e| Error::IoError(e))
+    }
+}
+
+impl<'a, T, const SIZE: usize> SbpParse<[T; SIZE]> for &'a [u8]
+where
+    &'a [u8]: SbpParse<T>,
+{
+    fn parse(&mut self) -> Result<[T; SIZE]> {
+        let mut vec = Vec::new();
+        for _ in 0..SIZE {
+            vec.push(self.parse()?);
+        }
+        vec.try_into().map_err(|_| crate::Error::ParseError)
+    }
```

Comment:
Not sure if there is a practical difference, but we could skip the intermediate vec here with something like

```rust
impl<'a, T, const SIZE: usize> SbpParse<[T; SIZE]> for &'a [u8]
where
    &'a [u8]: SbpParse<T>,
{
    fn parse(&mut self) -> Result<[T; SIZE]> {
        let mut data: [std::mem::MaybeUninit<T>; SIZE] =
            unsafe { std::mem::MaybeUninit::uninit().assume_init() };

        for elem in &mut data[..] {
            unsafe {
                std::ptr::write(elem.as_mut_ptr(), self.parse()?);
            }
        }

        let ptr = &mut data as *mut _ as *mut [T; SIZE];
        let res = unsafe { ptr.read() };
        core::mem::forget(data);

        Ok(res)
    }
}
```
Honestly not sure what the last few lines do, I wanted to write

```rust
impl<'a, T, const SIZE: usize> SbpParse<[T; SIZE]> for &'a [u8]
where
    &'a [u8]: SbpParse<T>,
{
    fn parse(&mut self) -> Result<[T; SIZE]> {
        let mut data: [std::mem::MaybeUninit<T>; SIZE] =
            unsafe { std::mem::MaybeUninit::uninit().assume_init() };

        for elem in &mut data[..] {
            unsafe {
                std::ptr::write(elem.as_mut_ptr(), self.parse()?);
            }
        }

        unsafe { Ok(std::mem::transmute::<_, [T; SIZE]>(data)) }
    }
}
```

But apparently you can't transmute with const generics atm. We could probably also avoid the unsafe by adding a `Default` trait bound so we could init the array like `let mut data = [Default::default(), SIZE];`

Not sure this is worth the unsafe, but seems like it would avoid putting some stuff on the heap?

Reply:
I was trying to figure out how to avoid the heap allocation, but couldn't figure out how to use `MaybeUninit` without using even more unstable nightly APIs. I think we should consider something like what you've recommended, all of the fixed size arrays are pretty small so a heap allocation seems overly expensive if it can be avoided. Thanks for the starting point! I'll try and work it in.

Reply:
Should have linked this earlier but for reference - https://doc.rust-lang.org/nomicon/unchecked-uninit.html
Explains `uninit().assume_init()` can work (I think)

Reply:
Also based on that, I think the loop is better off as -

```rust
for i in 0..SIZE {
    data[i] = std::mem::MaybeUninit::new(self.parse()?);
}
```

Reply:
I think what I have will leak memory if the array fills up partially then errors.`MaybeUninit` stops drop from getting called. I think that's fine if `T` is just a scalar - are there messages with nested arrays/vecs? Otherwise I think we need to manually drop whatever got created

Reply:
SBP can only nest fixed length arrays (and strings), an unbounded array can't be followed by anything else and has to be made up of fields with a known size. If we change the string type to also used fixed size arrays then we should be fine. That being said it seems like we're asking for trouble by not calling drop on objects when we should.